### PR TITLE
Separate objects necessary for extension/testing from core flytekit

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
This continues to hide the folder structure of flytekit by adding two new folders `extend` and `testing`.  That is,

* `from flytekit import <task, workflow, Email, etc.>` - This is what users will primarily interact with when writing workflows
* `from flytekit.testing import task_mock, patch` - These are the two functions that users can use for unit testing tasks and workflows that are not locally runnable.  (more detailed api docs to come, also it doesn't work on launch plans right now, we'll add that.)
* `from flytekit.extend import <Interface, Promise, TypeEngine, etc.>` - This is what contributors will use.  Please see the included `plugins/` in this repo for examples.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
* Rename `annotated` to `core`
* Make all the plugins use only things exposed through either the `flytekit`, `flytekit.extend`, or `flytekit.testing` subpackages.

### CODE REVIEW

File: plugins/awssagemaker/flytekitplugins/awssagemaker/training.py [hunk 1]
```diff
@@ -7,12 +7,17 @@
 from google.protobuf.json_format import MessageToDict
 
 import flytekit
-from flytekit.annotated.base_task import IgnoreOutputs, PythonTask, kwtypes
-from flytekit.annotated.context_manager import ExecutionState, FlyteContext, SerializationSettings
-from flytekit.annotated.interface import Interface
-from flytekit.annotated.python_function_task import PythonFunctionTask
-from flytekit.annotated.task import TaskPlugins
+from flytekit import FlyteContext, PythonFunctionTask
 from flytekit.common.tasks.sdk_runnable import ExecutionParameters
```
Comment (kumare3) [inline plugins/awssagemaker/flytekitplugins/awssagemaker/training.py:11]: I dont like this part that ExecutionParameters is coming from sdk_runnable. can we change that as well?

**Review by kumare3 — COMMENTED** [review]
Looks good to me, one comment. Now that you are changing this, might as well see if we can  expose ExecutionParameters

File: flytekit/extend/__init__.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from flytekit.common.translator import get_serializable
```
Comment (katrogan) [inline flytekit/extend/__init__.py:1]: Nice, will we document this?

Comment (wild-endeavor) [reply flytekit/extend/__init__.py:1]: not now.  but the goal is that anything that's in any of these three directories will have documentation as extensive as the `FlyteFile` one.

File: flytekit/core/testing.py [hunk 3]
```diff
@@ -2,9 +2,9 @@
 from typing import Union
 from unittest.mock import MagicMock
 
-from flytekit.annotated.base_task import PythonTask
-from flytekit.annotated.reference_entity import ReferenceEntity
-from flytekit.annotated.workflow import Workflow
+from flytekit.core.base_task import PythonTask
```
Comment (kumare3) [inline flytekit/core/testing.py:5]: i thought you want testing to be a separate module on its own, like `flytekit.testing` ?

Comment (wild-endeavor) [reply flytekit/core/testing.py:5]: yeah that's here.

Comment (wild-endeavor) [reply flytekit/core/testing.py:5]: so I know it's not a lot of code, but I left the actual implementation in `flytekit.core.testing` and i'm importing them in `flytekit/testing/__init__.py`.  Just for consistency.  I think it's okay?  bit circuitous but that's okay.

**Comment by wild-endeavor** [comment]
ExecutionParameters added

**Review by kumare3 — APPROVED** [review]

**Comment by honnix** [comment]
:+1:
